src/core/ext/transport/chttp2/transport/context_list.h

```diff
@@ -26,74 +26,43 @@
-  // within the current chunk.
-  void Append(void* context, size_t byte_offset_in_stream,
-              int64_t relative_start_pos_in_chunk,
-              int64_t num_traced_bytes_in_chunk);
 
-  bool IsEmpty() { return context_list_entries_.empty(); }
-
-  // Interprets the passed arg as a pointer to ContextList type and executes the
-  // function set using grpc_http2_set_write_timestamps_callback method with
-  // each context in the list and \a ts. It also deletes/frees up the passed
-  // ContextList after this operation. It is intended as a callback and hence
-  // does not take a ref on \a error. The fn receives individual contexts in the
-  // same order in which they were Appended.
-  static void Execute(void* arg, Timestamps* ts, grpc_error_handle error);
-
-  // Executes the passed function \a cb with each context in the list. The
-  // arguments provided to cb include the trace_context_, byte_offset_,
-  // traced_bytes_relative_start_pos_ and num_traced_bytes_ for each context in
-  // the context list. It also deletes/frees up the ContextList after this
-  // operation. The cb receives individual contexts in the same order in which
-  // they were Appended.
-  static void ForEachExecuteCallback(
-      ContextList* list,
-      absl::FunctionRef<void(void*, size_t, int64_t, int64_t)> cb);
+/// An RPC trace context and associated information. Each RPC/stream is
+/// associated with a unique \a context. A new ContextList entry is created when
+/// a chunk of data stored in an outgoing buffer is going to be
+// sent over the wire. A data chunk being written over the wire is multiplexed
+// with bytes from multiple RPCs. If one such RPC is traced, we store the
+// following information about the traced RPC:
+// - byte_offset_in_stream: Number of bytes belonging to that traced RPC which
+// have been sent so far from the start of the RPC stream.
+// - relative_start_pos_in_chunk: Starting offset of the traced RPC within
+// the current chunk that is being sent.
+// - num_traced_bytes_in_chunk: Number of bytes belonging to the traced RPC
+// within the current chunk.
+class ContextListEntry {
+ public:
+  ContextListEntry(void* context, int64_t relative_start_pos,
+                   int64_t num_traced_bytes, size_t byte_offset)
+      : trace_context_(context),
+        relative_start_pos_in_chunk_(relative_start_pos),
+        num_traced_bytes_in_chunk_(num_traced_bytes),
+        byte_offset_in_stream_(byte_offset) {}
 
-  // Use this function to create a new ContextList instead of creating it
-  // manually.
-  static ContextList* MakeNewContextList() { return new ContextList; }
+  void* TraceContext() { return trace_context_; }
+  int64_t RelativeStartPosInChunk() { return relative_start_pos_in_chunk_; }
+  int64_t NumTracedBytesInChunk() { return num_traced_bytes_in_chunk_; }
+  size_t ByteOffsetInStream() { return byte_offset_in_stream_; }
 
  private:
```

Comment:
Since you already refactored, can we explicitly delete the default ctor or make it private? i.e. ContextListEntry() = delete;

Reply:
Fixed